sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -100,6 +101,36 @@ class FileStreamSource(
 
   logInfo(s"maxFilesPerBatch = $maxFilesPerBatch, maxFileAgeMs = $maxFileAgeMs")
 
+  ensureNoOverlapBetweenSourceAndArchivePath()
+
+  private def ensureNoOverlapBetweenSourceAndArchivePath(): Unit = {
+    @tailrec
+    def removeGlob(path: Path): Path = {
+      if (path.getName.contains("*")) {
+        removeGlob(path.getParent)
+      } else {
+        path
+      }
+    }
+
+    sourceOptions.sourceArchiveDir match {
+      case None =>
+      case Some(archiveDir) =>
+        val sourceUri = removeGlob(qualifiedBasePath).toUri
+        val archiveUri = new Path(archiveDir).toUri
+
+        val sourcePath = sourceUri.getPath
+        val archivePath = archiveUri.getPath
```

Comment:
we need to use `fs.makeQualified` to turn all user provided paths to absolute paths as the user may just pass a relative path.